### PR TITLE
Correctly range check structs/unions when they are coming from a macro

### PR DESCRIPTION
**Pull request by gerazo**
Fixes: https://github.com/Ericsson/clang/issues/416
On tmux, there errors are fixed and tmux is checked without a problem.
On redis, the errors are all turned into "not a total ordering" problems. This change definitely allows more things to be imported, so there is a chance that more problems will come, still, it raises coverage.

### CODE REVIEW

**Review by balazske — APPROVED** [review]
Looks good, but it would be better to find a method for this check that does not involve comparing locations at all.

**Comment by gerazo** [comment]
Well, the point is to make sure that a declaration is in the body of a function. I don't know how to do this without comparing ranges, but maybe there is a way.
My other problem is that how robust is this solution. The record can come from a macro, the function can also come from that... It seems that getFileLoc is used for such thing at many places and inside it, it follows the expansion until it finds an isFileID loc. Still there may be better get***Loc functions for this.

File: unittests/AST/ASTImporterTest.cpp [hunk 1]
```diff
@@ -1014,6 +1014,31 @@ TEST_P(ASTImporterTestBase, ImportRecordDeclInFuncParams) {
...
+TEST_P(ASTImporterTestBase, ImportRecordDeclInFuncFromMacro) {
+  Decl *FromTU = getTuDecl(
+      "#define NONAME_SIZEOF(type) sizeof(struct{type *dummy;}) \n"
+      "int declToImport(){ return NONAME_SIZEOF(int); }",
+      Lang_C,
+      "input.c");
+  auto From = FirstDeclMatcher<FunctionDecl>().match(FromTU, functionDecl());
+  ASSERT_TRUE(From);
+  auto To = Import(From, Lang_C);
+  EXPECT_TRUE(To);
```
Comment (martong) [inline unittests/AST/ASTImporterTest.cpp]: Actually, the test could be more strict. In it's current form it does not fail on the baseline.
We could add this extra check:
```
  ASSERT_TRUE(To);
  EXPECT_TRUE(MatchVerifier<FunctionDecl>().match(
      To, functionDecl(hasDescendant(unaryExprOrTypeTraitExpr()))));
```

File: lib/AST/ASTImporter.cpp [hunk 2]
```diff
@@ -1104,16 +1104,23 @@ bool ASTNodeImporter::ImportDeclParts(NamedDecl *D, DeclContext *&DC,
   FunctionDecl *FunDecl;
   if (isa<RecordDecl>(D) && (FunDecl = dyn_cast<FunctionDecl>(OrigDC)) &&
       FunDecl->hasBody()) {
-    SourceRange RecR = D->getSourceRange();
-    SourceRange BodyR = FunDecl->getBody()->getSourceRange();
+    SourceManager &SM = D->getASTContext().getSourceManager();
```
Comment (martong) [inline lib/AST/ASTImporter.cpp]: I was thinking about an alternative solution which does not involve source locations. Go through all the params and if any of them has an equal type to the currently imported record then bail out:
```
  if (isa<RecordDecl>(D) && (FunDecl = dyn_cast<FunctionDecl>(OrigDC)) &&
      FunDecl->hasBody()) {
    for (const ParmVarDecl *P : FunDecl->parameters()) {
      auto getLeafPointeeType = [](const Type *T) {
        while (T->isPointerType() || T->isArrayType()) {
          T = T->getPointeeOrArrayElementType();
        }
        return T;
      };
      const Type *LeafT =
          getLeafPointeeType(P->getType().getCanonicalType().getTypePtr());
      if (auto * RT = dyn_cast<RecordType>(LeafT)) {
        if (RT->getDecl() == D) {
          Importer.FromDiag(D->getLocation(), diag::err_unsupported_ast_node)
              << D->getDeclKindName();
          Importer.setEncounteredUnsupportedNode(true);
          return true;
        }
      }
    }
  }
```

Also we could have more tests for pointee/pointer stuff if we choose this, something like:
```
TEST_P(ASTImporterTestBase, ImportRecordDeclInFuncParams) {
  // This construct is not supported by ASTImporter.
  Decl *FromTU = getTuDecl(
      "int declToImport(struct data_t{int a;int b;} d){ return 0; }",
      Lang_C,
      "input.c");
  auto From = FirstDeclMatcher<FunctionDecl>().match(FromTU, functionDecl());
  ASSERT_TRUE(From);
  auto To = Import(From, Lang_C);
  EXPECT_EQ(To, nullptr);
}

TEST_P(ASTImporterTestBase, ImportRecordDeclInFuncParams2) {
  // This construct is not supported by ASTImporter.
  Decl *FromTU = getTuDecl(
      "int declToImport(struct data_t{int a;int b;} ***d[3]){ return 0; }",
      Lang_C,
      "input.c");
  auto From = FirstDeclMatcher<FunctionDecl>().match(FromTU, functionDecl());
  ASSERT_TRUE(From);
  auto To = Import(From, Lang_C);
  EXPECT_EQ(To, nullptr);
}
```

**Comment by martong** [comment]
There is one other think which is not entirely clean. How was this connected to "total ordering"?

**Comment by gerazo** [comment]
The total ordering problem is a different bug currently here: https://github.com/Ericsson/clang/issues/428 probably needs rephrasing. There is no connection. In fact, the total ordering problem comes only late when all the checkers found problems and those are about to be reported. So if you have any crash earlier, than the crash supresses the total ordering problem.

File: unittests/AST/ASTImporterTest.cpp (hunk 1 above)
Comment (gerazo) [reply unittests/AST/ASTImporterTest.cpp]: You're right, will do.

**Comment by gerazo** [comment]
Check this out now @balazske and @martong . Tests are running perfectly, tmux and redis gives the exact same result. I would say, it is identical to the previous solution.

**Comment by martong** [comment]
LGTM, just could you please squash the first two commits into one? (I think it is still worth to preserve the source range version in a commit).

File: unittests/AST/ASTImporterTest.cpp [hunk 3]
```diff
@@ -1005,7 +1005,46 @@ TEST_P(ASTImporterTestBase, ImportRecordTypeInFunc) {
...
+  EXPECT_EQ(To, nullptr);
+}
+
+TEST_P(ASTImporterTestBase, ImportRecordDeclInFuncParams2) {
+  // This construct is not supported by ASTImporter.
+  Decl *FromTU = getTuDecl(
+      "int declToImport(struct data_t{int a;int b;} ***d){ return 0; }",
+      Lang_C,
+      "input.c");
+  auto From = FirstDeclMatcher<FunctionDecl>().match(FromTU, functionDecl());
```
Comment (balazske) [inline unittests/AST/ASTImporterTest.cpp]: Probably use `auto *` at such places (not only here)?

File: unittests/AST/ASTImporterTest.cpp [hunk 4]
```diff
@@ -1005,7 +1005,46 @@ TEST_P(ASTImporterTestBase, ImportRecordTypeInFunc) {
 TEST_P(ASTImporterTestBase, ImportRecordDeclInFuncParams) {
   // This construct is not supported by ASTImporter.
   Decl *FromTU = getTuDecl(
-      "int declToImport(struct data_t{int a;int b;} *d){ return 0; }",
+      "int declToImport(struct data_t{int a;int b;} d){ return 0; }",
+      Lang_C,
+      "input.c");
+  auto From = FirstDeclMatcher<FunctionDecl>().match(FromTU, functionDecl());
```
Comment (balazske) [inline unittests/AST/ASTImporterTest.cpp]: Probably it is better to use `functionDecl(hasName("declToImport"))` to make sure not something "hidden" (implicit) function is found.

Comment (gerazo) [reply unittests/AST/ASTImporterTest.cpp]: Thanks, I am applying these suggestions.

**Comment by gerazo** [comment]
@martong There are 4 commits, the first 2 should be squashed? The history will have the newer state anyway, no? Probably I don't get the purpose.

**Comment by martong** [comment]
@gerazo,
Doesn't really matter, just would be good to preserve the SourceRange based version somehow in the history.

**Comment by martong** [comment]
Actually, the first commit out of the four commits keep the SourceLocation based version. So, let's just squash the last 3 commit into one. Does it make sense?

**Comment by gerazo** [comment]
Ok, I did it.

**Comment by martong** [comment]
Thanks Zozo, nice work! :)

**Comment by gerazo** [comment]
Thanks. Can I start the upstreaming?

**Comment by martong** [comment]
Of course :)

**Comment by gerazo** [comment]
Review is open: https://reviews.llvm.org/D49792

**Pull request merged**